### PR TITLE
Add scenarios! macro for automatic feature discovery

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- add `scenarios!` macro to auto-generate tests for every Scenario in a directory
- document macro usage and record design decisions
- mark roadmap item for boilerplate reduction complete

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make markdownlint`
- `make nixie` *(fails: Module not found '/tmp/bunx-0-@mermaid-js/mermaid-cli@latest/node_modules/.old-9E1E015DE0C71792/install.mjs')*

------
https://chatgpt.com/codex/tasks/task_e_689fad0a8ef48322b4732108a205a0ae

## Summary by Sourcery

Implement the `scenarios!` macro for automatic feature discovery, update documentation to reflect its usage, and add tests to verify its behavior.

New Features:
- Add `scenarios!` procedural macro to scan feature file directories and auto-generate a test function for each Scenario

Documentation:
- Document `scenarios!` macro usage in the user guide, update the design proposal, and mark the boilerplate reduction roadmap item complete

Tests:
- Add feature files and behavioural tests to validate the `scenarios!` macro output

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

Bumpy Road Ahead
crates/rstest-bdd-macros/src/macros/scenarios.rs: scenarios

What lead to degradation?

scenarios has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is one single, nested block per function
Why does this problem occur?

A Bumpy Road is a function that contains multiple chunks of nested conditional logic inside the same function. The deeper the nesting and the more bumps, the lower the code health.

A bumpy code road represents a lack of encapsulation which becomes an obstacle to comprehension. In imperative languages there’s also an increased risk for feature entanglement, which leads to complex state management. CodeScene considers the following rules for the code health impact: 1) The deeper the nested conditional logic of each bump, the higher the tax on our working memory. 2) The more bumps inside a function, the more expensive it is to refactor as each bump represents a missing abstraction. 3) The larger each bump – that is, the more lines of code it spans – the harder it is to build up a mental model of the function. The nesting depth for what is considered a bump is levels of conditionals.
How to fix it?

Bumpy Road implementations indicate a lack of encapsulation. Check out the detailed description of the [Bumpy Road code health issue](https://codescene.com/blog/bumpy-road-code-complexity-in-context/).

A Bumpy Road often suggests that the function/method does too many things. The first refactoring step is to identify the different possible responsibilities of the function. Consider extracting those responsibilities into smaller, cohesive, and well-named functions. The [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html) refactoring is the primary response.

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        let entry = entry?;
+        let path = entry.path();
+        if path.is_dir() {
+            files.extend(collect_feature_files(&path)?);
+        } else if path.extension().is_some_and(|e| e == "feature") {
+            files.push(path);
+        }
+    }
+    Ok(files)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs]: Has this now been resolved?

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    let mut used_names = HashSet::new();
+
+    for abs_path in feature_paths {
+        let rel_path = abs_path
+            .strip_prefix(&manifest_dir)
+            .map_or_else(|_| abs_path.clone(), Path::to_path_buf);
+
+        let feature = match parse_and_load_feature(rel_path.as_path()) {
+            Ok(f) => f,
+            Err(err) => return err,
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs]: Has this now been resolved?

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                }
+            }
+        } else if file_type.is_dir() {
+            files.extend(collect_feature_files(&path)?);
+        } else if file_type.is_file() && path.extension().is_some_and(|e| e == "feature") {
+            files.push(path);
+        }
+    }
+    Ok(files)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:63]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:63]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

Complex Method
crates/rstest-bdd-macros/src/macros/scenarios.rs: collect_feature_files

What lead to degradation?

collect_feature_files has a cyclomatic complexity of 9, threshold = 9
Why does this problem occur?

A Complex Method has a high cyclomatic complexity. The recommended threshold for the Rust language is a cyclomatic complexity lower than 9.
How to fix it?

There are many reasons for Complex Method. Sometimes, another design approach is beneficial such as a) modeling state using an explicit state machine rather than conditionals, or b) using table lookup rather than long chains of logic. In other scenarios, the function can be split using [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html). Just make sure you extract natural and cohesive functions. Complex Methods can also be addressed by identifying complex conditional expressions and then using the [DECOMPOSE CONDITIONAL](https://refactoring.com/catalog/decomposeConditional.html) refactoring.
Helpful refactoring examples

To get a general understanding of what this code health issue looks like - and how it might be addressed - we have prepared some diffs for illustrative purposes.
SAMPLE
complex_method.js
function postItem(item) {
  if (!item.id) {
    if (item.x != null && item.y != null) {
      post(item);
    } else {
      throw Error("Item must have x and y");
    }
    // extract a separate function for creating new item
    postNew(item);
  } else {
    if (item.x < 10 && item.y > 25) {
      put(item);
    } else {
      throw Error("Item must have an x and y value between 10 and 25");
    }
    // and one for updating existing items
    updateItem(item);
  }
}
function postNew(item) {
  validateNew(item);
  post(item);
}
function updateItem(item) {
  validateUpdate(item);
  put(item);
}

File: docs/users-guide.md [hunk 4]
```diff
@@ -168,6 +168,26 @@ with `rstest` features such as parameterized tests and asynchronous fixtures.
...
+use rstest_bdd_macros::{given, then, when, scenarios};
+
+#[given("a precondition")] fn precondition() {}
+#[when("an action occurs")] fn action() {}
+#[then("events are recorded")] fn events() {}
+
+scenarios!("tests/features/auto");
+```
+
+Generated tests cannot currently accept fixtures; use `#[scenario]` when
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:188]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- Consider replacing manual recursion in collect_feature_files with the walkdir crate for more robust directory traversal and symlink support.
- Add a test case for duplicate scenario titles to verify that the deduplication suffix logic produces unique function names as intended.
- Consider using env!("CARGO_MANIFEST_DIR") for compile-time manifest directory resolution instead of std::env::var to avoid runtime macro expansion failures.

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            if let Ok(target) = fs::metadata(&path) {
+                let target_type = target.file_type();
+                if target_type.is_file() && path.extension().is_some_and(|e| e == "feature") {
+                    files.push(path);
+                }
+            }
+        } else if file_type.is_dir() {
+            files.extend(collect_feature_files(&path)?);
+        } else if file_type.is_file() && path.extension().is_some_and(|e| e == "feature") {
+            files.push(path);
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs]: Has this now been resolved?

File: crates/rstest-bdd-macros/src/macros/scenarios.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs]: Has this now been resolved?

File: crates/rstest-bdd-macros/tests/scenarios_macro.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+#[given("a precondition")]
+fn precondition() {}
+
+#[when("an action occurs")]
+fn action() {}
+
+#[then("events are recorded")]
+fn events_recorded() {}
+
+scenarios!("tests/features/auto");
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/tests/scenarios_macro.rs]: Has this now been resolved?

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        feature_path,
+        scenario_name: data.name,
+        steps: data.steps,
+        examples: data.examples,
+    };
+    Ok(TokenStream2::from(generate_scenario_code(
+        config,
+        std::iter::empty(),
+    )))
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:122]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: crates/rstest-bdd-macros/tests/trybuild.rs [hunk 8]
```diff
@@ -4,6 +4,8 @@
 fn step_macros_compile() {
     let t = trybuild::TestCases::new();
     t.pass("tests/fixtures/step_macros.rs");
+    // `scenarios!` should succeed when the directory exists.
+    // t.pass("tests/fixtures/scenarios_autodiscovery.rs");
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/tests/trybuild.rs:8]: @coderabbitai Please create a new GitHub issue for this

File: docs/users-guide.md [hunk 9]
```diff
@@ -168,6 +168,28 @@ with `rstest` features such as parameterized tests and asynchronous fixtures.
...
 
+## Autodiscovering scenarios
+
+For large suites, it is tedious to bind each scenario manually. The
+`scenarios!` macro scans a directory recursively for `.feature` files and
+generates a module with a test for every `Scenario` found. Each test is named
+after the feature file and scenario title. Identifiers are sanitised
+(ASCII-only) and deduplicated by appending a numeric suffix when collisions
+occur.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:179]: Per `docs/lille-physics-and-world-engine-roadmap.md`:

**Velocity and Acceleration**:

-  Implement `terminal_velocity` clamping for unsupported entities subject to gravity.

Ensure that the DBSP circuit is the source of truth for any inferred behaviour in the game world.

Ensure that the new feature is validated with unit tests using `rstest` and behavioural tests using `rust-rspec`, covering happy and unhappy paths and any relevant edge cases. Record any design decisions taken in the design document.

See, `docs/lille-physics-engine-design.md`, `docs/bevy-headless-testing.md`, `docs/rust-testing-with-rstest-fixtures.md`, and `docs/testing-declarative-game-logic-in-dbsp.md`

On completion of this feature, mark the relevant entry in the roadmap as "done".

File: crates/rstest-bdd-macros/tests/features/auto/beta.feature [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+  Scenario: two – with symbols & spaces
+    Given a precondition
+    When an action occurs
+    Then events are recorded
+
+  Scenario: one
+    Given a precondition  
+    When an action occurs
+    Then events are recorded
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/tests/features/auto/beta.feature:15]: @coderabbitai Please create a new GitHub issue for this

File: docs/users-guide.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:179]: @coderabbitai Please create a new GitHub issue for this

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+        counter += 1;
+    }
+    used.insert(name.clone());
+    name
+}
+
+fn generate_scenario_test(
+    ctx: &ScenarioTestContext<'_>,
+    used_names: &mut HashSet<String>,
+) -> Result<TokenStream2, TokenStream> {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:95]: @coderabbitai Please create a new GitHub issue for this

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+/// Generate the test for a single scenario within a feature.
+/// Context for generating a scenario test.
+struct ScenarioTestContext<'a> {
+    feature: &'a Feature,
+    scenario_idx: usize,
+    feature_stem: &'a str,
+    manifest_dir: &'a Path,
+    rel_path: &'a Path,
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:79]: @coderabbitai Please create a new GitHub issue for this

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+            Err(e) => {
+                let err_str = e.to_string();
+                let io_err = e
+                    .into_io_error()
+                    .unwrap_or_else(|| io::Error::other(err_str));
+                Some(Err(io_err))
+            }
+        })
+        .collect::<Result<_, _>>()?;
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:66]: @coderabbitai Please create a new GitHub issue for this

File: crates/rstest-bdd-macros/src/macros/scenarios.rs [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+/// Recursively collect all `.feature` files under `base`.
+fn collect_feature_files(base: &Path) -> std::io::Result<Vec<PathBuf>> {
+    use std::io;
+    use walkdir::WalkDir;
+
+    fn is_feature_file(path: &Path) -> bool {
+        path.extension()
+            .and_then(|e| e.to_str())
+            .is_some_and(|ext| ext.eq_ignore_ascii_case("feature"))
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/macros/scenarios.rs:47]: @coderabbitai Please create a new GitHub issue for this

File: crates/rstest-bdd-macros/src/lib.rs [hunk 15]
```diff
@@ -27,3 +27,34 @@ pub fn then(attr: TokenStream, item: TokenStream) -> TokenStream {
...
+/// test per Gherkin `Scenario`.
+///
+/// Path semantics:
+/// - The `dir` argument must be a string literal.
+/// - It is resolved relative to `CARGO_MANIFEST_DIR` at macro-expansion time.
+///
+/// Expansion:
+/// - Emits a module named after `dir` (sanitised) containing one test function
+///   per discovered scenario.
+/// - Each generated test executes the matched steps via the registered
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-macros/src/lib.rs:41]: @coderabbitai Please create a new GitHub issue for this

File: docs/users-guide.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:179]: @coderabbitai Please create a new GitHub issue for this
